wp-content/themes/pub/wporg-learn-2024/inc/block-config.php

```diff
@@ -557,3 +559,74 @@ function inject_other_filters( $key ) {
 		printf( '<input type="hidden" name="s" value="%s" />', esc_attr( $wp_query->query['s'] ) );
 	}
 }
+
+/**
+ * Modifies the courses query for the My Courses page, by adding courses to be excluded.
+ *
+ * @param array $parsed_block The parsed block data.
+ * @return array The modified block data.
+ */
+function filter_student_course_list( $parsed_block ) {
+	if (
+		'core/query' !== $parsed_block['blockName'] ||
+		'course' !== ( $parsed_block['attrs']['query']['postType'] ?? '' ) ||
+		! isset( $parsed_block['attrs']['queryId'] ) ||
+		! get_the_ID() === Sensei()->settings->get_my_courses_page_id()
+	) {
+		return $parsed_block;
+	}
+
+	if ( ! array_key_exists( 'exclude', $parsed_block['attrs']['query'] ) || ! is_array( $parsed_block['attrs']['query']['exclude'] ) ) {
+		$parsed_block['attrs']['query']['exclude'] = array();
+	}
+
+	$filter_param_key = get_student_course_filter_query_var_name();
+	$selected_option  = isset( $_GET[ $filter_param_key ] ) ? sanitize_text_field( wp_unslash( $_GET[ $filter_param_key ] ) ) : '';
+
+	// The courses query with active and completed statuses have already been filtered in Sensei LMS, and can correctly display the course lists.
+	// See https://github.com/Automattic/sensei/blob/trunk/includes/blocks/course-list/class-sensei-course-list-student-course-filter.php#L114-L123.
+	if ( ! empty( $selected_option ) && ( 'active' === $selected_option || 'completed' === $selected_option ) ) {
+		return $parsed_block;
+	}
+
+	$exclude_ids = get_course_ids_to_be_excluded();
+
+	$parsed_block['attrs']['query']['exclude'] = array_merge(
+		$parsed_block['attrs']['query']['exclude'],
+		$exclude_ids
+	);
+
+	return $parsed_block;
+}
+
+/**
+ * Get a list of course IDs to be excluded from the course lists filtered by the 'all' status
+ * or those that do not belong to the student course options.
+ *
+ * Corresponds to https://github.com/Automattic/sensei/blob/trunk/includes/blocks/course-list/class-sensei-course-list-student-course-filter.php#L95
+ */
+function get_course_ids_to_be_excluded() {
+	$user_id = get_current_user_id();
+
+	if ( empty( $user_id ) ) {
+		return array();
+	}
+
+	$args = array(
+		'posts_per_page' => -1,
+		'fields'         => 'ids',
+	);
+
+	$learner_manager     = Sensei_Learner::instance();
+	$courses_query       = $learner_manager->get_enrolled_courses_query( $user_id, $args );
+	$included_course_ids = $courses_query->posts;
```

Comment:
We actually know the ids for the list of Courses here, and don't need to do another query and diff to work out what should be excluded from the query. I tried setting the `post__in` property of the block query to this array, rather than setting the `excludes` property but it didn't work. Could be worth looking further into this though.

Reply:
483e865
Ah, forgot that there's `post__in`. This makes the logic much simpler. Big thanks!